### PR TITLE
pkg/ansible/controller,pkg/helm/controller; Enable status subresource by default

### LINKED ISSUE #809: Enable the status subresource for k8s 1.12
## Feature Request

The status subresource for CRDs is beta and enabled by default for k8s 1.12 clusters
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#subresources

Once the SDK updates to k8s 1.12 it should support enabling the status subresource for CRDs by default when generating the CRD yaml manifest.
See for more details: https://blog.openshift.com/kubernetes-custom-resources-grow-up-in-v1-10/

Along with that the examples using the client should be updated to use the status client `c.Status().Update()` to only update the status when needed.
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/interfaces.go#L88

With the status subresource enabled controllers can update the status, and ignore any events that don't have changes to `metadata.Generation` (only updates when spec changes) by filtering them out via predicates.

### PR DESCRIPTION
**Pull request by dymurray**
**Description of the change:**
Enable use of status subresource by default in operator-sdk.

*Ansible:*
Adds support for hybrid managed status between AO and user's Ansible. Get's latest version of resource to ensure that we are updating the latest version of the resource.

**Motivation for the change:**
In Kubernetes v1.11.0 the status subresource is enabled by default and so this PR enables the status subresource by default in operator-sdk.

*Ansible:*
Hybrid use case panics when user updates status from Ansible task.

Closes #809

### CODE REVIEW

**Comment by shawn-hurley** [comment]
@dymurray Before we merge this we need to make sure that we add some migration steps to some sort of doc to describe what it looks like if you're using a CRD without sub-resources and what to do to fix it.

**Comment by shawn-hurley** [comment]
/hold 

this would require v1.12 of kubernetes by default. @fabianvf @dymurray we may need to rethink this, my apolgies I forgot about this caveat when discussing this.

Lets plan on talking at some point today in person to determine what to do.

**Comment by dymurray** [comment]
Just correcting Shawn's comment, this requires v1.11 of Kubernetes by default:
https://github.com/kubernetes/kubernetes/pull/63598

**Comment by hasbro17** [comment]
fixes: https://github.com/operator-framework/operator-sdk/issues/809 as we're turning on the status subresource by default which is beta in k8s 1.11 per kubernetes/kubernetes#63598
Although I would like to see a manual confirmation of the status subresource being on by default (and not behind a feature gate) on a k8s 1.11 cluster.

**Comment by joelanford** [comment]
@dymurray Have you looked into the Helm error in this build at all? I'm seeing the same error trying to add a status subresource to a CRD.

 https://travis-ci.org/operator-framework/operator-sdk/builds/464000402?utm_source=github_status&utm_medium=notification

**Comment by joelanford** [comment]
I think it's a [bug](https://github.com/kubernetes/kubernetes/issues/67235) fixed in 1.11.3 [here](https://github.com/kubernetes/kubernetes/pull/66249), but I'm not positive. With 1.11.2 the helm e2e test works if I change the `operator-sdk new` command to `--kind=Memcached`. With 1.11.3, the helm e2e test works as is.

**Comment by dymurray** [comment]
@joelanford I haven't dug into it just yet but your link looks promising. I will report back if I find anything. It only began happening when I added the status writer to the helm controller.

File: doc/user/client.md [hunk 1]
```diff
@@ -244,6 +244,41 @@ func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, e
 }
 ```
 
+##### Updating Status Subresource
+When updating the status subresource from the client, the StatusWriter must be
```
Comment (hasbro17) [inline doc/user/client.md]:
```suggestion

When updating the status subresource from the client, the StatusWriter must be
```
Leave a new line between the section header and section content.

File: doc/user/client.md [hunk 2]
```diff
@@ -244,6 +244,41 @@ func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, e
...
+used which can be gotten with `Status()`
+##### Status
```
Comment (hasbro17) [inline doc/user/client.md]:
```suggestion

##### Status
```
Newline before new section header.

File: doc/user/client.md [hunk 3]
```diff
@@ -244,6 +244,41 @@ func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, e
...
+```Go
```
Comment (hasbro17) [inline doc/user/client.md]:
```suggestion

```Go
```
Newline after section header.

**Comment by hasbro17** [comment]
@dymurray Can you please update the PR title and description to more clearly say that we're enabling the status subresource by default.

**Comment by dymurray** [comment]
Helm CI tests are related to the bug Joe mentioned above. I ran the tests with `kubectl` version `1.11.3` and the error went away.

File: Gopkg.toml [hunk 4]
```diff
@@ -8,23 +8,23 @@
 
 [[override]]
   name = "k8s.io/api"
-  version = "kubernetes-1.11.2"
+  version = "kubernetes-1.11.3"
```
Comment (joelanford) [inline Gopkg.toml]: Should we go ahead and bump up to 1.11.5?

Also, should we update the `pkg/scaffold/gopkgtoml.go` with these same deps?

Comment (dymurray) [reply Gopkg.toml]: No opposition here to bumping to 1.11.5. And yes good catch I will update the scaffold as well.

File: Gopkg.toml [hunk 5]
```diff
@@ -8,23 +8,23 @@
...
 
 [[override]]
   name = "k8s.io/apimachinery"
-  version = "kubernetes-1.11.2"
+  version = "kubernetes-1.11.5"
 
 [[override]]
   name = "k8s.io/apiextensions-apiserver"
-  version = "kubernetes-1.11.2"
+  version = "kubernetes-1.11.5"
```
Comment (LiliC) [inline Gopkg.toml]: There is no mention on why we bumped this to 1.11.5?

File: doc/user/client.md [hunk 6]
```diff
@@ -244,6 +244,44 @@ func (r *ReconcileApp) Reconcile(request reconcile.Request) (reconcile.Result, e
 }
 ```
 
+##### Updating Status Subresource
+
+When updating the status subresource from the client, the StatusWriter must be
```
Comment (LiliC) [inline doc/user/client.md]: Can we link to this document here https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource wdyt?

Comment (joelanford) [reply doc/user/client.md]: Also, might want to highlight in our doc that using `r.Client.Update(ctx, mem)` will NOT update the `status`. I ran across that trying to set status conditions and adding/removing finalizers, which requires two separate updates.

**Comment by dymurray** [comment]
I have removed and client library bump from this PR because of #807

File: doc/user/client.md (hunk 6 above)
Comment (dymurray) [reply doc/user/client.md]: Done, thanks guys!

File: Gopkg.toml (hunk 5 above)
Comment (dymurray) [reply Gopkg.toml]: No longer bumping this thanks to #807

File: pkg/ansible/controller/reconcile.go [hunk 7]
```diff
@@ -189,15 +190,23 @@ func (r *AnsibleOperatorReconciler) Reconcile(request reconcile.Request) (reconc
...
-func (r *AnsibleOperatorReconciler) markRunning(u *unstructured.Unstructured) error {
+func (r *AnsibleOperatorReconciler) markRunning(u *unstructured.Unstructured, namespacedName types.NamespacedName) error {
+	// Get the latest resource to prevent updating a stale status
+	err := r.Client.Get(context.TODO(), namespacedName, u)
+	if apierrors.IsNotFound(err) {
+		return err
+	}
+	if err != nil {
+		return err
+	}
```
Comment (hasbro17) [inline pkg/ansible/controller/reconcile.go]: Isn't this:
```Go
	if apierrors.IsNotFound(err) {
		return err
	}
	if err != nil {
		return err
	}
```
equivalent to:
```Go
	if err != nil {
		return err
	}
```

Comment (dymurray) [reply pkg/ansible/controller/reconcile.go]: Thanks for catching that, initially had a log statement there.

**Comment by dymurray** [comment]
Thanks @hasbro17, added the note to the changelog.

File: CHANGELOG.md [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 - The SDK now uses logr as the default logger to unify the logging output with the controller-runtime logs. Users can still use a logger of their own choice. See the [logging doc](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/logging.md) on how the SDK initializes and uses logr.
 - Ansible Operator CR status better aligns with [conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#typical-status-properties). ([#639](https://github.com/operator-framework/operator-sdk/pull/639))
+- The SDK now generates the CRD with the status subresource enabled by default. See the [client doc](https://github.com/operator-framework/operator-sdk/blob/masster/doc/user/client.md#updating-status-subresource) on how to update the status ssubresource.
```
Comment (hasbro17) [inline CHANGELOG.md]: `ssubresource` ==> `subresource`
Can you also add the link to this PR #787  like we do for other changes.

Comment (dymurray) [reply CHANGELOG.md]: Done, thanks.

**Comment by hasbro17** [comment]
My bad, I forgot to squash before merging.

**Pull request merged**